google/data_source_google_kms_secret.go

```diff
@@ -0,0 +1,67 @@
+package google
+
+import (
+	"google.golang.org/api/cloudkms/v1"
+
+	"encoding/base64"
+	"fmt"
+	"github.com/hashicorp/terraform/helper/schema"
+	"log"
+	"time"
+)
+
+func dataSourceGoogleKmsSecret() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceGoogleKmsSecretRead,
+		Schema: map[string]*schema.Schema{
+			"crypto_key": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"ciphertext": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"plaintext": &schema.Schema{
+				Type:      schema.TypeString,
+				Computed:  true,
+				Sensitive: true,
+			},
+		},
+	}
+}
+
+func dataSourceGoogleKmsSecretRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	cryptoKeyId, err := parseKmsCryptoKeyId(d.Get("crypto_key").(string), config)
+
+	if err != nil {
+		return err
+	}
+
+	ciphertext := d.Get("ciphertext").(string)
+
+	kmsDecryptRequest := &cloudkms.DecryptRequest{
+		Ciphertext: ciphertext,
+	}
+
+	decryptResponse, err := config.clientKms.Projects.Locations.KeyRings.CryptoKeys.Decrypt(cryptoKeyId.cryptoKeyId(), kmsDecryptRequest).Do()
+
+	if err != nil {
+		return fmt.Errorf("Error decrypting ciphertext: %s", err)
+	}
+
+	plaintext, err := base64.StdEncoding.DecodeString(decryptResponse.Plaintext)
```

Comment:
Shouldn't these two calls be the other way around - ie, base64 decode the ciphertext first, and then send the ciphertext to KMS?

Reply:
Nope.  `decrypt` expects the encrypted data that was originally returned from `encrypt`, which is already in base64, and it returns some plaintext that is base64 encoded.

[Docs](https://cloud.google.com/kms/docs/reference/rest/v1/projects.locations.keyRings.cryptoKeys/decrypt)

Reply:
Looks on closer inspection like I got confused by a difference between how the gcloud CLI handles this (ie, it spits raw bytes out to the terminal) and how the REST API does (ie, it base64 encodes everything). Sorry about that.

Reply:
No worries!